src/engine/sys/sdl_glimp.cpp

```diff
@@ -558,196 +774,542 @@ static rserr_t GLimp_SetMode( int mode, bool fullscreen, bool noborder )
+		{ 3, 3, coreProfile, false },
+		{ 3, 2, coreProfile, true },
+		{ 3, 1, compatibilityProfile, false },
+		{ 3, 0, compatibilityProfile, false },
+		{ 2, 1, compatibilityProfile, true },
+		{ 2, 0, compatibilityProfile, true },
+		{ 1, 5, compatibilityProfile, true },
+		{ 1, 4, compatibilityProfile, true },
+		{ 1, 3, compatibilityProfile, true },
+		{ 1, 2, compatibilityProfile, true },
+		{ 1, 1, compatibilityProfile, true },
+		{ 1, 0, compatibilityProfile, true },
+	};
+
+	logger.Debug( "Validating best OpenGL context." );
+
+	for ( int colorBits : {24, 16} )
 	{
-		if ( glContext != nullptr )
+		// Test core profiles, stop on first valid one.
+		for ( auto& row : glSupportArray )
 		{
-			SDL_GL_DeleteContext( glContext );
-			glContext = nullptr;
+			if ( GLEWmajor < 2 )
+			{
+				// GLEW version < 2.0.0 doesn't support OpenGL core profiles.
+				break;
+			}
+
+			if ( !r_glExtendedValidation->integer && !row[ 3 ] )
+			{
+				continue;
+			}
+
+			glConfiguration testConfiguration = {};
+
+			testConfiguration.major = row[ 0 ];
+			testConfiguration.minor = row[ 1 ];
+			testConfiguration.profile = row[ 2 ];
+			testConfiguration.colorBits = colorBits;
+
+			if ( testConfiguration.profile == compatibilityProfile )
+			{
+				break;
+			}
+
+			if ( !GLimp_RecreateWindowWhenChange( false, false, testConfiguration ) )
+			{
+				return rserr_t::RSERR_INVALID_MODE;
+			}
+
+			if ( GLimp_CreateContext( testConfiguration ) )
+			{
+				bestValidatedConfiguration = testConfiguration;
+				return rserr_t::RSERR_OK;
+			}
 		}
 
-		if ( window != nullptr )
+		// Test compatibility profiles if no core profile is valid.
+		for ( auto& row : glSupportArray )
 		{
-			SDL_GetWindowPosition( window, &x, &y );
-			logger.Debug("Existing window at %dx%d before being destroyed", x, y );
-			SDL_DestroyWindow( window );
-			window = nullptr;
+			if ( !r_glExtendedValidation->integer )
+			{
+				if ( !row[ 3 ] || row[ 2 ] == coreProfile )
+				{
+					continue;
+				}
+			}
+
+			glConfiguration testConfiguration = {};
+
+			testConfiguration.major = row[ 0 ];
+			testConfiguration.minor = row[ 1 ];
+			testConfiguration.profile = compatibilityProfile;
+			testConfiguration.colorBits = colorBits;
+
+			if ( !GLimp_RecreateWindowWhenChange( false, false, testConfiguration ) )
+			{
+				return rserr_t::RSERR_INVALID_MODE;
+			}
+
+			if ( GLimp_CreateContext( testConfiguration ) )
+			{
+				bestValidatedConfiguration = testConfiguration;
+				return rserr_t::RSERR_OK;
+			}
 		}
-		// we come back here if we couldn't get a visual and there's
-		// something we can switch off
+	}
+
+	if ( bestValidatedConfiguration.major == 0 )
+	{
+
+		return rserr_t::RSERR_MISSING_GL;
+	}
+
+	return rserr_t::RSERR_OLD_GL;
+}
+
+static rserr_t GLimp_ApplyCustomOptions( const int GLEWmajor, const bool fullscreen, const bool bordered, const glConfiguration &bestConfiguration, glConfiguration &requestedConfiguration, bool &customOptions )
```

Comment:
It seems better for this function to only return a new configuration struct. The caller can do window/context creation calls.

Also can you just use the `operator!=` between the bestConfiguration and the customized options to get the answer to whether custom options were applied?